### DISCUSSION TITLE
Virtual entities custom join with variable

### DISCUSSION THREAD

**Discussion by veselin-angelov** [mikro-orm/mikro-orm#4970] [General]
**Is your feature request related to a problem? Please describe.**
I want to create a virtual entity in which I can join entities on custom condition with a variable
```ts
@Entity({
  expression: (em: EntityManager, where, options: FindOptions<any, any>) => {
    return em
      .createQueryBuilder(EntityA, 'a')
      .leftJoin('a.entityBPivot', 'bPivot', {
        'bPivot.entityD': entityDId, // entityDId is a variable
      })
  },
})
export class VirtualEntity {}
```

**Describe the solution you'd like**
Be able to pass customOptions to FindOptions, so it would look like this:
```ts
@Entity({
  expression: (em: EntityManager, where, options: FindOptions<any, any>) => {
    return em
      .createQueryBuilder(EntityA, 'a')
      .leftJoin('a.entityBPivot', 'bPivot', {
        'bPivot.entityD': options.customOptions['entityDId'],
      })
  },
})
export class VirtualEntity {}
```

**Proposed solution**
```ts
export interface FindOptions<T, P extends string = never, F extends string = never> {
  ...
  customOptions?: Object;
}
```

This would allow you to pass anything you like to the virtual entity and use it as you find fit.

**Comment by B4nan** [comment]
The proposal seems a bit confusing to me, I am not sure I want to have something like that in the `FindOptions` mainly because it would work only for virtual entities (and also because the name sounds pretty weird, options inside options...).

Technically, you can already do this with the help of type assertions, you will get the full options object in the expression callback, there is nothing to change in the code itself for this to work.

I will think about this a bit more in the evening.

**Reply by veselin-angelov** [reply]
yeah, I just realized that I can just cast it. My idea is to have some property where you can pass whatever data you want to use in the virtual entity expression callback.
